mlir/test/Dialect/Vector/vector-transfer-flatten.mlir

```diff
@@ -471,25 +471,27 @@ func.func @regression_non_contiguous_dim_read(%subview : memref<1x3x3x2xf32, str
 }
 
 //       CHECK:  #[[$MAP:.+]] = affine_map<()[s0] -> (s0 * 2)>
-// CHECK-LABEL:    func.func @regression_non_contiguous_dim_read(
-//       CHECK:      %[[COLLAPSE:.+]] = memref.collapse_shape %{{.*}} {{\[}}[0], [1], [2, 3]] : memref<1x3x3x2xf32, strided<[40, 10, 2, 1], offset: ?>> into memref<1x3x6xf32, strided<[40, 10, 1], offset: ?>>
-//       CHECK:     %[[APPLY:.*]] = affine.apply #[[$MAP]]()
+// CHECK-LABEL:  func.func @regression_non_contiguous_dim_read(
+//       CHECK:    %[[COLLAPSE:.+]] = memref.collapse_shape %{{.*}} {{\[}}[0], [1], [2, 3]] : memref<1x3x3x2xf32, strided<[40, 10, 2, 1], offset: ?>> into memref<1x3x6xf32, strided<[40, 10, 1], offset: ?>>
+//       CHECK:    %[[APPLY:.*]] = affine.apply #[[$MAP]]()
 
 // CHECK-128B-LABEL: func @regression_non_contiguous_dim_read(
 //       CHECK-128B:   memref.collapse_shape
 
 // -----
 
-func.func @unsupported_non_contiguous_dim_write(%value : vector<2x2xf32>,
+func.func @regression_non_contiguous_dim_write(%value : vector<2x2xf32>,
```

Comment:
This turns an existing testcase that couldn't be collapsed into one that can be collapsed. Is any other testcase left testing remaining cases that more fundamentally can't be collapsed?

Also, is any testcase testing `in_bounds = false` and specifically any edge case where the ability to collapse depends on `in_bounds` ?

Reply:
> E.g., there is a CHECK-128B:   memref.collapse_shape in this and other testcases.

But `CHECK-128B` tests a specific configuration of the pass rather than the generic version, right? 

>  Fundamentally, we can not collapse three inner dimensions because the strides is [x, 10, 2, 1] but not [x, 6, 2, 1]

Just to double check - this is referring to `regression_non_contiguous_dim_write`? In which collapsing _does_ happen?

In general, I think that it would be good to identify and document all edge cases through tests (most of the stuff is already here). I would start by renaming this test:

> @regression_non_contiguous_dim_write

"Regression" doesn't really help differentiate this test from other tests. It's a missed opportunity to document what makes this test case unique - i.e. non-zero indices :) (*) So, perhaps:

* `@transfer_write_non_contigous_dims_mismatch_non_zero_idx`

Next, you could move it near `@transfer_write_dims_mismatch_non_contiguous` - both tests seem to check sth very similar, two edge cases for `vector.transfer_write`. I find it very helpful when such edge cases are clustered together.

> > Also, is any testcase testing in_bounds = false and specifically any edge case where the ability to collapse depends on in_bounds ?
>
> There are no negative tests for the case. I can add one to iterate it.

I've been actually discussing this with someone recently :) We don't need it for this PR, but it would be good to add a TODO somewhere (I can look into this).

(*) In general, I'm not a fan of regression tests, but appreciate that others find the valuable.

Reply:
>> Fundamentally, we can not collapse three inner dimensions because the strides is [x, 10, 2, 1] but not [x, 6, 2, 1]

> Just to double check - this is referring to regression_non_contiguous_dim_write? In which collapsing does happen?

Yes, it is referring to the test. However, I was wrong on the statement because it is 2D vector. I will add a 3D vector case, and check that it is not collapsed.

> "Regression" doesn't really help differentiate this test from other tests. It's a missed opportunity to document what makes this test case unique - i.e. non-zero indices :) (*) So, perhaps:
>     @transfer_write_non_contigous_dims_mismatch_non_zero_idx
> Next, you could move it near @transfer_write_dims_mismatch_non_contiguous - both tests seem to check sth very similar, two edge cases for vector.transfer_write. I find it very helpful when such edge cases are clustered together.

I also found "regression" is weird, but I'd like to follow the naming style like the above xfer_read test. The new name is great, and I'll send a follow-up about naming and structure of the file.

> I've been actually discussing this with someone recently :) We don't need it for this PR, but it would be good to add a TODO somewhere (I can look into this).

Wait, I thought that I added the tests! Then I found that it is on my local file.. my bad. adding now

Reply:
I just found that we already have a negative test. It does not matter whether the last index is zero or not. So I'm going to merge the change and send a follow-up for the naming and structure.

https://github.com/llvm/llvm-project/blob/0e743ecca018e9fa7482dc8bec62eae8cd0c929b/mlir/test/Dialect/Vector/vector-transfer-flatten.mlir#L215-L221